packages/core/src/sync-historical/index.ts

```diff
@@ -464,613 +344,595 @@ export const createHistoricalSync = (
+          syncStore.insertBlocks({ blocks: [block], chainId: args.chain.id }),
+          syncStore.insertTransactions({
+            transactions,
+            chainId: args.chain.id,
+          }),
+          syncStore.insertTransactionReceipts({
+            transactionReceipts,
+            chainId: args.chain.id,
+          }),
+          syncStore.insertTraces({
+            traces: traces.map((trace) => ({
+              trace,
+              block: block!,
+              transaction: transactionsByHash.get(trace.transactionHash)!,
+            })),
+            chainId: args.chain.id,
+          }),
+          syncStore.insertLogs({ logs, chainId: args.chain.id }),
+        ]);
+      };
 
-            case "trace":
-            case "transfer": {
-              await syncTraceOrTransferFilter(
-                filter as TraceFilter | TransferFilter,
-                interval,
-                context,
-              );
-              break;
-            }
-          }
+      let blockCount = 0;
+      let transactionCount = 0;
+      let receiptCount = 0;
+      let traceCount = 0;
 
-          await blockPromise;
-        }),
+      // Same memory usage as `sync-realtime`.
+      const MAX_BLOCKS_IN_MEM = Math.max(
+        args.chain.finalityBlockCount * 2,
+        100,
       );
 
-      const blocks = await Promise.all(blockCache.values());
-
-      await Promise.all([
-        args.syncStore.insertBlocks(
-          { blocks, chainId: args.chain.id },
-          context,
-        ),
-        args.syncStore.insertTransactions(
-          {
-            transactions: blocks.flatMap((block) =>
-              block.transactions.filter(({ hash }) =>
-                transactionsCache.has(hash),
-              ),
-            ),
-            chainId: args.chain.id,
-          },
-          context,
-        ),
-        ...Array.from(childAddressesCache.entries()).map(
-          ([factory, childAddresses]) =>
-            args.syncStore.insertChildAddresses(
-              {
-                factory,
-                childAddresses,
-                chainId: args.chain.id,
-              },
-              context,
-            ),
-        ),
-      ]);
-
-      // Add corresponding intervals to the sync-store
-      // Note: this should happen after so the database doesn't become corrupted
-      if (args.chain.disableCache === false) {
-        await args.syncStore.insertIntervals(
-          {
-            intervals: intervalsToSync,
-            chainId: args.chain.id,
-          },
-          context,
-        );
-      }
+      const queue = createQueue({
+        browser: false,
+        initialStart: true,
+        concurrency: MAX_BLOCKS_IN_MEM,
+        worker: syncBlockData,
+      });
 
-      let childAddressCount = 0;
-      for (const childAddresses of childAddressesCache.values()) {
-        childAddressCount += childAddresses.size;
-      }
+      await Promise.all(intervalRange(interval).map(queue.add));
```

Comment:
This could in practice enqueue a large number of no-op tasks, correct? Perhaps we should have a hard cap on the `interval` size.